src/app/templates/SwapForm/SwapForm.tsx

```diff
@@ -198,23 +170,30 @@ export const SwapForm: FC = () => {
 
     formAnalytics.trackSubmit(analyticsProperties);
 
+    if (!fromRoute3Token || !toRoute3Token || !inputValue.amount) {
+      return;
+    }
+
     try {
       setOperation(undefined);
-      const routingFeeOpParams = await getRoutingFeeTransferParams(
-        bestTradeWithSlippageTolerance,
-        feeAmount,
-        account.publicKeyHash,
-        tezos
+
+      const route3SwapOpParams = await getRoute3SwapOpParams(
+        fromRoute3Token,
+        toRoute3Token,
+        inputValue.amount,
+        slippageRatio
       );
-      const tradeOpParams = await getTradeOpParams(
-        bestTradeWithSlippageTolerance,
-        account.publicKeyHash,
+
+      const { approve, revoke } = await getTransferPermissions(
```

Comment:
maybe move approve & revoke logic into `useRoute3` hook?  
let it prepare all swap params